spec/unit/coinbase/smart_contract_spec.rb

```diff
@@ -198,6 +198,388 @@
+
+    before do
+      allow(smart_contracts_api).to receive(:read_contract)
+    end
+
+    describe 'network parameter' do
+      it 'uses the correct normalized network ID' do
+        result
+        expect(smart_contracts_api).to have_received(:read_contract).with('base-sepolia', anything, anything)
+      end
+    end
+
+    describe 'address parameter' do
+      it 'uses the provided contract address' do
+        result
+        expect(smart_contracts_api).to have_received(:read_contract).with(anything, contract_address, anything)
+      end
+    end
+
+    describe 'method parameter' do
+      it 'includes the method name in the request params' do
+        result
+        expect(smart_contracts_api).to have_received(:read_contract).with(anything, anything,
+                                                                          hash_including(method: method_name))
+      end
+    end
+
+    describe 'abi parameter' do
+      context 'when provided' do
+        let(:abi) { [{ 'name' => 'testMethod', 'inputs' => [], 'outputs' => [] }] }
+
+        it 'sends JSON encoded hash as abi in request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(anything, anything,
+                                                                            hash_including(abi: abi.to_json))
+        end
+      end
+
+      context 'when explicitly set to nil' do
+        let(:abi) { nil }
+
+        it 'omits the abi from request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(
+            anything,
+            anything,
+            hash_including(abi: nil)
+          )
+        end
+      end
+
+      context 'when omitted' do
+        subject(:result) do
+          described_class.read(
+            network: network,
+            contract_address: contract_address,
+            method: method_name
+          )
+        end
+
+        it 'omits the abi from request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(
+            anything,
+            anything,
+            hash_including(abi: nil)
+          )
+        end
+      end
+    end
+
+    describe 'args parameter' do
+      context 'when provided' do
+        let(:args) { { 'value' => 123 } }
+
+        it 'sends JSON encoded hash as args in request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(anything, anything,
+                                                                            hash_including(args: args.to_json))
+        end
+      end
+
+      context 'when explicitly set to nil' do
+        let(:args) { nil }
+
+        it 'sends null as args in request params' do
+          result
+          expect(smart_contracts_api).to have_received(:read_contract).with(
+            anything,
+            anything,
+            hash_including(args: 'null')
```

Comment:
We probably want this to pass the same value when `omitted` or when explicitly `nil`?

Reply:
yup good catch!